memory/src/main/java/org/opensearch/ml/memory/index/InteractionsIndex.java

```diff
@@ -340,4 +346,89 @@ public void deleteConversation(String conversationId, ActionListener<Boolean> li
+     * boolean query around the original query, AND a term query over conversation id
+     * @param conversationId the id of the conversation to query over
+     * @param request the original search request
+     * @param listener receives the search response from this query
+     */
+    public void searchInteractions(String conversationId, SearchRequest request, ActionListener<SearchResponse> listener) {
+        conversationMetaIndex.checkAccess(conversationId, ActionListener.wrap(access -> {
+            if (access) {
+                try (ThreadContext.StoredContext threadContext = client.threadPool().getThreadContext().stashContext()) {
+                    ActionListener<SearchResponse> internalListener = ActionListener.runBefore(listener, () -> threadContext.restore());
+                    request.indices(indexName);
+                    QueryBuilder originalQuery = request.source().query();
+                    BoolQueryBuilder newQuery = new BoolQueryBuilder();
+                    newQuery.must(originalQuery);
+                    newQuery.must(new TermQueryBuilder(ConversationalIndexConstants.INTERACTIONS_CONVERSATION_ID_FIELD, conversationId));
+                    request.source().query(newQuery);
+                    client.admin().indices().refresh(Requests.refreshRequest(indexName), ActionListener.wrap(refreshResponse -> {
+                        client.search(request, internalListener);
+                    }, e -> {
+                        log.error("Failed to refresh interactions index during search interactions ", e);
+                        internalListener.onFailure(e);
+                    }));
+                } catch (Exception e) {
+                    listener.onFailure(e);
+                }
+            } else {
+                String userstr = client
+                    .threadPool()
+                    .getThreadContext()
+                    .getTransient(ConfigConstants.OPENSEARCH_SECURITY_USER_INFO_THREAD_CONTEXT);
+                String user = User.parse(userstr) == null ? ActionConstants.DEFAULT_USERNAME_FOR_ERRORS : User.parse(userstr).getName();
+                throw new OpenSearchSecurityException("User [" + user + "] does not have access to conversation " + conversationId);
+            }
+        }, e -> { listener.onFailure(e); }));
+    }
+
+    /**
+     * Get a single interaction
+     * @param conversationId id of the conversation this interaction belongs to
+     * @param interactionId id of this interaction
+     * @param listener receives the interaction
+     */
+    public void getInteraction(String conversationId, String interactionId, ActionListener<Interaction> listener) {
+        if (!clusterService.state().metadata().hasIndex(indexName)) {
+            listener.onFailure(new IndexNotFoundException("cannot get interaction since the interactions index does not exist", indexName));
+            return;
+        }
```

Comment:
Wouldn't OS throw this error anyway when you execute the Get request below?

Reply:
sure, but this way I shortcut the access control request